platform/generic/xuantie/xuantie_ipmc_test.c

```diff
@@ -0,0 +1,228 @@
+
+	/* Clean and invalidate core's dcache */
+	asm volatile(XUANTIE_DCACHE_INVAL_CLEAN_ALL);
+
+	/* Close the Dcache */
+	csr_clear(THEAD_C9XX_CSR_MHCR, 0x1 << 1);
+	asm volatile(XUANTIE_SYNC_IS);
+
+	/* Close the snoop */
+	csr_clear(THEAD_C9XX_CSR_MSMPR, 0x1);
+
+	csr_set(CSR_MIE, MIP_MTIP | MIP_STIP | MIP_SEIP | MIP_MEIP);
+
+	wfi();
+
+	__asm__ __volatile__(
+		"jalr x0, %0\n\t"
+		:
+		: "r"(zsb_addr)
+		: "memory");
+}
+
+static int xuantie_hart_suspend(u32 suspend_type)
+{
+	u32 hartid = current_hartid();
+
+	/* Use the generic code for retentive suspend. */
+	if (!(suspend_type & SBI_HSM_SUSP_NON_RET_BIT))
+		return SBI_ENOTSUPP;
+
+	xuantie_hart_poweroff(hartid, true);
+
+	return 0;
+}
+
+static void xuantie_hart_resume(void)
+{
+}
+
+int xuantie_hart_start(u32 hartid, ulong saddr)
+{
+	return 0;
+}
+
+int xuantie_hart_stop(void)
+{
+	u32 hartid = current_hartid();
+
+	xuantie_hart_poweroff(hartid, false);
+
+	return 0;
+}
+
+static const struct sbi_hsm_device xuantie_hsm = {
+	.name			= "xuantie-hsm",
+	.hart_suspend	= xuantie_hart_suspend,
+	.hart_resume	= xuantie_hart_resume,
+	.hart_start 	= xuantie_hart_start,
+	.hart_stop		= xuantie_hart_stop,
+};
+
+static void xuantie_enable_ddr_auto_selfrefresh()
+{
+}
+
+static void xuantie_system_clk_suspend()
+{
+}
+
+static void xuantie_system_power_suspend()
+{
+}
+
+static int xuantie_system_suspend_check(u32 sleep_type)
+{
+	return sleep_type == SBI_SUSP_SLEEP_TYPE_SUSPEND ? 0 : SBI_EINVAL;
+}
+
+static int xuantie_system_suspend(u32 sleep_type, unsigned long mmode_resume_addr)
+{
+	u32 hartid = current_hartid();
+
+	if (sleep_type != SBI_SUSP_SLEEP_TYPE_SUSPEND)
+		return SBI_EINVAL;
+
+	/* enable ddr auto selfrefresh */
+	xuantie_enable_ddr_auto_selfrefresh();
+
+	/* enable wakeup irqs, disable other irqs, Linux framwork already done */
+	/* There just make sure the mask of irq is disabled */
+	/* xuantie_set_wakeup_source(); */
+
+	/* xuantie_system_clk_suspend , include clock system and plls*/
+	xuantie_system_clk_suspend();
+
+	/* xuantie_system_power_suspend , close the module powers which have no wakeup sources */
+	xuantie_system_power_suspend();
+
+	/* enable cpuidle, cpu poweroff after wfi */
+	xuantie_hart_poweroff(hartid, true);
+
+	return 0;
+}
+
+static struct sbi_system_suspend_device xuantie_system_suspend_dev = {
+	.name = "xuantie-system-suspend",
+	.system_suspend_check = xuantie_system_suspend_check,
+	.system_suspend = xuantie_system_suspend,
+};
+
+static int xuantie_system_reset_check(u32 type, u32 reason)
+{
+	switch (type) {
+	case SBI_SRST_RESET_TYPE_SHUTDOWN:
+		return 1;
+	case SBI_SRST_RESET_TYPE_COLD_REBOOT:
+	case SBI_SRST_RESET_TYPE_WARM_REBOOT:
+		return 255;
+	}
+
+	return 0;
+}
+
+static void xuantie_system_reset(u32 type, u32 reason)
+{
+}
+
+static struct sbi_system_reset_device xuantie_reset = {
+	.name = "xuantie-reset",
+	.system_reset_check = xuantie_system_reset_check,
+	.system_reset = xuantie_system_reset
```

Comment:
NULL?

Reply:
Cannot be NULL. After registering the device, system_reset will be called directly.